c7n/resources/sns.py

```diff
@@ -464,6 +464,37 @@ class resource_type(TypeInfo):
             'TopicArn'
         )
 
+@SNSSubscription.filter_registry.register('unused')
+class UnusedSNSSubscription(CrossAccountAccessFilter):
+
+    """
+    Filters subscriptons based on invalid topic arn
+
+    :example:
+
+    .. code-block:: yaml
+
+            policies:
+              - name: sns-subscription-unused
+                resource: sns-subscription
+                filters:
+                  - type: unused
+    """
+
+    schema = type_schema('unused')
+    permissions = ('sns:GetTopicAttributes',)
+
+    def process(self, resources, event=None):
+        results = []
+        client = local_session(self.manager.session_factory).client('sns')
+        for r in resources:
+            try:
+                client.get_topic_attributes(TopicArn=r['TopicArn'])
```

Comment:
If we go the related resource route, we'll fetch SNS topics once (or reuse cached resources) rather than having to list one topic at a time this way.

This does mean that we won't be able to see topics outside the active account/region, but as far as I can tell we wouldn't be able to reliably use `get_topic_attributes` to cover cross-account/cross-region cases either.

Reply:
Hmmm... this is a good point @ajkerrigan . Is there a way to filter out subscriptions whose topics are in a different account and/or region? It would allow us to clean up 95+% of the zombie subscriptions without breaking a valid subscription to a topic that lives elsewhere.

Reply:
Thinking for this case maybe we could put a value filter first, to focus on subscriptions in the active account/region:

```yaml
filters:
  - type: value
    key: TopicArn
    op: glob
    value: "arn:aws:sns:{region}:{account_id}:*"
```

And then put this `unused` filter (given the changes, probably more accurate to call it a `topic` filter) after it. So this might be a reasonable policy for the filter's docstring example and/or unit test:

```yaml
filters:
  - type: value
    key: TopicArn
    op: glob
    value: "arn:aws:sns:{region}:{account_id}:*"
  - type: topic
    key: TopicArn
    value: absent
actions:
  - delete
```

An alternative would be for this filter to pass the TopicArn through for subscriptions outside the active account/region, so it would only be absent for known orphaned subscriptions. In a way that seems a bit friendlier to the policy author, but it also feels a bit "magical" :thinking:

Reply:
Yes @ajkerrigan I tested with current unused code and it works fine with above policy.  Try to write the test case and will push it sooner.